Implement internal means to consume user activation via test-driver.js

marcoscaceres opened the pull request:
#### 438bd44b426edf09e02f977d7a0fa30a9410f76e
<pre>
Implement internal means to consume user activation via test-driver.js
<a href="https://bugs.webkit.org/show_bug.cgi?id=247968">https://bugs.webkit.org/show_bug.cgi?id=247968</a>
rdar://102399664

This patch adds two APIs for testing internally within WebKit:

 * The Internals interface gains the ability to consume the user activation
   of a Window via internal.consumeTransientActivation().

 * And test_driver.js gains the ability to consume the user activation of
   a Window via test_driver.consume_user_activation(). This can be used
   for Web Platform Tests.

Collectively, these are useful for testing APIs that depend on transient activation,
as it allows us to easily check that the API works as expected if the user activation
has been consumed.

Reviewed by Chris Dumez.

* LayoutTests/fast/frames/consume_transient_activation-expected.txt: Added.
* LayoutTests/fast/frames/consume_transient_activation.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-same-and-cross-origin.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-same-and-cross-origin.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-keyboard-enter.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-keyboard-escape.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-mouse-left.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-mouse-right.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-pointerevent.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-crossorigin.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-crossorigin.sub.tentative-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-crossorigin.sub-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-crossorigin.sub.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-sameorigin.tentative-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-sameorigin-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-sameorigin.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/resources/child-two.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/resources/consumption-crossorigin-child.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/resources/consumption-sameorigin-child.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/resources/utils.js:
(getEvent):
(async consumeTransientActivation): Deleted.
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver.js:
(window.test_driver.consume_user_activation):
(window.test_driver_internal.async consume_user_activation):
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::consumeTransientActivation):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/256916@main">https://commits.webkit.org/256916@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8f9a76ee9aa8cc4e99e5c26a72fff7539c7ac0e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97171 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30310 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106689 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166961 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101142 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6713 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35172 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89561 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103379 "Hash c8f9a76e for PR 6539 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102838 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5047 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83797 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32024 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86896 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88723 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74951 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/460 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/444 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21649 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4769 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5242 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44148 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1684 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40965 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/fast/frames/consume_transient_activation.html:
```diff
@@ -0,0 +1,71 @@
+        var x = frame.offsetLeft + frame.offsetWidth / 2;
+        var y = frame.offsetTop + frame.offsetHeight / 2;
+
+        eventSender.mouseMoveTo(x, y);
```
We should probably be using activateAt() from ui-helper.js so that the test can be cross-platform.

cdumez commented on LayoutTests/fast/frames/consume_transient_activation.html:
```diff
@@ -0,0 +1,70 @@
+      var innerFrame2;
+
+      function dispatchUserGesture() {
+        var x = frame.offsetLeft + frame.offsetWidth / 2;
```
We normally use 4-space indentation in WebKit:
https://webkit.org/code-style-guidelines/#indentation-4-spaces

marcoscaceres committed:
Implement internal means to consume user activation via test-driver.js
https://bugs.webkit.org/show_bug.cgi?id=247968
rdar://102399664

This patch adds two APIs for testing internally within WebKit:

 * The Internals interface gains the ability to consume the user activation
   of a Window via internal.consumeTransientActivation().

 * And test_driver.js gains the ability to consume the user activation of
   a Window via test_driver.consume_user_activation(). This can be used
   for Web Platform Tests.

Collectively, these are useful for testing APIs that depend on transient activation,
as it allows us to easily check that the API works as expected if the user activation
has been consumed.

Reviewed by Chris Dumez.

* LayoutTests/fast/frames/consume_transient_activation-expected.txt: Added.
* LayoutTests/fast/frames/consume_transient_activation.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-same-and-cross-origin.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-same-and-cross-origin.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-keyboard-enter.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-keyboard-escape.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-mouse-left.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-mouse-right.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/activation-trigger-pointerevent.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-crossorigin.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-crossorigin.sub.tentative-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-crossorigin.sub-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-crossorigin.sub.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-sameorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-sameorigin.tentative-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-sameorigin-expected.txt.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/consumption-sameorigin.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/resources/child-two.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/resources/consumption-crossorigin-child.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/resources/consumption-sameorigin-child.html:
* LayoutTests/imported/w3c/web-platform-tests/html/user-activation/resources/utils.js:
(getEvent):
(async consumeTransientActivation): Deleted.
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver.js:
(window.test_driver.consume_user_activation):
(window.test_driver_internal.async consume_user_activation):
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::consumeTransientActivation):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/256916@main
